### PR TITLE
Generate exam refactor

### PR DESCRIPTION
**Pull request by MorAbo**
refactoring generate exam with @OfekN11 and @MayElnathan

### CODE REVIEW

File: src/main/business/applicationFacade.js [hunk 1]
```diff
@@ -394,6 +394,39 @@ class ApplicationFacade{
...
+        return {
+            stem: data.question.stem,
+            appendix: data.question.appendix,
+            key: { text: "fake answer", explanation: "fake explanation"},
+            distractors: [
+                { text: "fake answer 1", explanation: "fake explanation"},
+                { text: "fake answer 2", explanation: "fake explanation"},
+                { text: "fake answer 3", explanation: "fake explanation"},
+                { text: "fake answer 4", explanation: "fake explanation"}
+            ]
```
Comment (MorAbo) [inline src/main/business/applicationFacade.js:427]: @OfekN11 @mayer-goldberg  please notice this server request

File: src/main/business/applicationFacade.js [hunk 2]
```diff
@@ -366,6 +366,20 @@ class ApplicationFacade{
...
     }
 
+    /**
+     * return a list of meta question relevant to add to the new exam:
+     * - each question has at least 1 key and 4 distractors that are not used yet in the exam
+     * - the keys and distractors returned are not used yet in the exam
+     * @throws {Error} - If the user is not signed in or does not have the necessary permissions
+     * @return {MetaQuestion[]} all the relevant question in a meta question structure
+     */
+    getMetaQuestionsForExam(data) {
```
Comment (MorAbo) [inline src/main/business/applicationFacade.js:376]: @OfekN11 @MayElnathan

File: src/main/business/applicationFacade.js [hunk 3]
```diff
@@ -394,6 +408,37 @@ class ApplicationFacade{
...
+            distractors: [
+                { text: "fake answer 1", explanation: "fake explanation"},
+                { text: "fake answer 2", explanation: "fake explanation"},
+                { text: "fake answer 3", explanation: "fake explanation"},
+                { text: "fake answer 4", explanation: "fake explanation"}
+            ]
+        };
+    }
+
+    removeQuestionFromExam(data) {
```
Comment (MorAbo) [inline src/main/business/applicationFacade.js:437]: @OfekN11 @MayElnathan

File: ReactApp/src/pages/generate-exam.js [hunk 4]
```diff
@@ -31,24 +29,20 @@ const Page = () => {
...
-
-    setCurrentQuestion(null)
+  }, [setQuestions]);
+
+  const addQuestion = async (question, AutomaticGenerateState) => {
+    try{
+      console.log(question)
+      let request = AutomaticGenerateState ? serverPath.ADD_AUTOMATIC_META_QUESTION : serverPath.ADD_MANUAL_META_QUESTION
+      const { examQuestion } = await requestServer(request, httpsMethod.POST, { question });
+      console.log(examQuestion)
```
Comment (OfekN11) [inline ReactApp/src/pages/generate-exam.js:43]: same

File: ReactApp/src/sections/create-exam/add-question-to-exam-popup.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  const [selectedDistractors, setSelectedDistractors] = useState([]);
+  const [generateState, setGenerateState] = useState(false); // Use boolean for simplicity
+
+  const handleSaveQuestion = () => {
+    addQuestion({
+      stem: selectedMetaQuestion.stem,
+      appendix: selectedMetaQuestion.appendix,
+      key: selectedKey,
+      distractors: selectedDistractors
+    }, generateState);
```
Comment (OfekN11) [inline ReactApp/src/sections/create-exam/add-question-to-exam-popup.js:22]: Please pass the id as well.

I think a bettwe approch should be 

addQuestion{...selectedMetaQuestion, key:selectedKey, distractors:selectdDistractors, ect....}

Please just make sure to put the '...selectedMetaQuestion' at the beginning (due to js behaviour)

File: src/main/server.js [hunk 6]
```diff
@@ -248,12 +306,16 @@ function createServer(options) {
     //       answer: str,
     //       distractors: [str]
     //     }]
+    //     numVersions: num,
+    //     examReason: str
```
Comment (MorAbo) [inline src/main/server.js:310]: @OfekN11 @MayElnathan

**Review by OfekN11 — COMMENTED** [review]
Great job!
You did more than you have should for the server part, thanks!

@RoiTief If you do a cr please ignore server and service issues, I will change and fix all necessary things with different PRs, but it is great that I have the doc and her expactation from the system...

File: ReactApp/src/utils/rest-api-call.js
Comment (OfekN11) [inline ReactApp/src/utils/rest-api-call.js:1]: Please add the expected return types to pathToReturnTypeMap (a const in this file)
That will help us prevent wasting a lot of time on debugging

**Review by OfekN11 — CHANGES_REQUESTED** [review]

File: src/main/business/applicationFacade.js [hunk 7]
```diff
@@ -310,6 +310,20 @@ class ApplicationFacade{
...
     }
 
+    /**
+     * return a list of meta question relevant to add to the new exam:
+     * - each question has at least 1 key and 4 distractors that are not used yet in the exam
+     * - the keys and distractors returned are not used yet in the exam
+     * @throws {Error} - If the user is not signed in or does not have the necessary permissions
+     * @return {MetaQuestion[]} all the relevant question in a meta question structure
+     */
+    async getMetaQuestionsForExam(data) {
```
Comment (RoiTief) [inline src/main/business/applicationFacade.js:320]: add a call to `validateParameters` so we know what is the received data
maybe @OfekN11 can do it for you

File: src/main/business/applicationFacade.js [hunk 8]
```diff
@@ -336,6 +350,45 @@ class ApplicationFacade{
         return this.userController.updateUser(data);
     }
 
+
+    addManualMetaQuestionToExam(data) {
```
Comment (RoiTief) [inline src/main/business/applicationFacade.js:354]: add a call to `validateParameters` so we know what is the received data
maybe @OfekN11 can do it for you

File: src/main/business/applicationFacade.js [hunk 9]
```diff
@@ -336,6 +350,45 @@ class ApplicationFacade{
...
+        return {
+            id: data.question.selectedMetaQuestion.id,
+            stem: data.question.selectedMetaQuestion.stem,
+            appendix: data.question.selectedMetaQuestion.appendix,
+            key: data.question.selectedKey,
+            distractors: data.question.selectedDistractors
+        };
+    }
+
+    addAutomaticQuestionToExam(data) {
```
Comment (RoiTief) [inline src/main/business/applicationFacade.js:367]: add a call to `validateParameters` so we know what is the received data
maybe @OfekN11 can do it for you

File: src/main/server.js [hunk 10]
```diff
@@ -216,6 +216,56 @@ function createServer(options) {
...
+    //       id: num
+    //       stem: str,
+    //       key: {answer: str, explanation: str },
+    //       distractors: [{answer: str, explanation: str }],
+    //      appendix: {
+    //          title: str,
+    //          tag: str,
+    //          content: str
+    //       }
+    //     }
```
Comment (RoiTief) [inline src/main/server.js:234]: having this documentation here isn't ideal as this area isn't really viewed/worked with.
moving it to the facade would be a better choice.
and you can document only the returned value, and have the sent data be self documented with the call to `validateParameters`

File: src/main/server.js [hunk 11]
```diff
@@ -216,6 +216,56 @@ function createServer(options) {
...
```
Comment (RoiTief) [inline src/main/server.js:253]: having this documentation here isn't ideal as this area isn't really viewed/worked with.
moving it to the facade would be a better choice
and you can document only the returned value, and have the sent data be self documented with the call to `validateParameters`

File: src/main/server.js [hunk 12]
```diff
@@ -216,6 +216,56 @@ function createServer(options) {
...
```
Comment (RoiTief) [inline src/main/server.js:266]: having this documentation here isn't ideal as this area isn't really viewed/worked with.
moving it to the facade would be a better choice
and you can document only the returned value, and have the sent data be self documented with the call to `validateParameters`

File: src/main/server.js (hunk 6 above)
Comment (RoiTief) [reply src/main/server.js:304]: having this documentation here isn't ideal as this area isn't really viewed/worked with.
moving it to the facade would be a better choice
and you can document only the returned value, and have the sent data be self documented with the call to validateParameters

File: src/main/service.js [hunk 13]
```diff
@@ -484,9 +575,13 @@ module.exports = {
     viewAllUsers: viewAllUsers,
     deleteUser: deleteUser,
     getAllMetaQuestions: getAllMetaQuestions,
+    getMetaQuestionsForExam,
```
Comment (RoiTief) [inline src/main/service.js:578]: keep the already implemented standard of `key: value` even when they are the same

**Comment by RoiTief** [comment]
image
can we align 'Created Questions` to the left? and add in brackets the number of questions in the exam. 

image
A search by keyword is way stronger, can we have that as well?

File: src/main/business/applicationFacade.js (hunk 7 above)
Comment (OfekN11) [reply src/main/business/applicationFacade.js:320]: This should be (and will be) done at the business level, at the implementation.

Thats due to the possibilty that more function will call it.

The validation should be done by the ones who use the data.

File: src/main/business/applicationFacade.js (hunk 8 above)
Comment (OfekN11) [reply src/main/business/applicationFacade.js:354]: This should be (and will be) done at the business level, at the implementation.

Thats due to the possibilty that more function will call it.

The validation should be done by the ones who use the data.

File: src/main/business/applicationFacade.js (hunk 9 above)
Comment (OfekN11) [reply src/main/business/applicationFacade.js:367]: This should be (and will be) done at the business level, at the implementation.

Thats due to the possibilty that more function will call it.

The validation should be done by the ones who use the data.

File: src/main/server.js (hunk 10 above)
Comment (OfekN11) [reply src/main/server.js:234]: Agree, will be moved

File: src/main/server.js (hunk 11 above)
Comment (OfekN11) [reply src/main/server.js:253]: Agree, will be moved

File: src/main/server.js (hunk 12 above)
Comment (OfekN11) [reply src/main/server.js:266]: Agree, will be moved

File: src/main/service.js (hunk 13 above)
Comment (OfekN11) [reply src/main/service.js:578]: Cant agree that continue bad habits is the right thing to do it. but who cares...

File: src/main/server.js (hunk 6 above)
Comment (OfekN11) [reply src/main/server.js:304]: Agreed, will be moved

**Comment by MorAbo** [comment]
> image can we align 'Created Questions` to the left? and add in brackets the number of questions in the exam.
> 
> image A search by keyword is way stronger, can we have that as well?

i will not align it to the left. i dont think it looks good. but i will add the number of question in the exam and a search by keyword

**Pull request merged**